v3/src/components/graph/components/graph-component.tsx

```diff
@@ -35,14 +35,19 @@ export const GraphComponent = observer(({broker, v2Document}: IProps) => {
   const enableAnimation = useRef(true)
   const dataset = broker?.selectedDataSet || broker?.last
   const dotsRef = useRef<SVGSVGElement>(null)
+  const graphController = useRef<GraphController>()
+
+  const getNewGraphController = () => {
+    return new GraphController({
+      graphModel: defaultGraphModel,
+      dataset, layout, enableAnimation, instanceId, dotsRef, v2Document
+    })
+  }
+
+  useEffect(() => {
+    graphController.current = getNewGraphController()
+  },[dataset, layout, instanceId, v2Document])
```

Comment:
I suggest moving all of this into a custom hook, `useGraphController`, which takes as arguments the things it needs from this function (`graphModel`, `enableAnimation`, `dotsRef`, `v2Document`), recovers others from context where possible (`dataset`, `layout`, `instanceId`), and then returns the `graphControllerRef`.

Reply:
@kswenson in the latest iteration I moved things into a hook, but was only able to get `instanceId` from context without introducing new errors.

Reply:
I added a commit which implements what I had in mind.